File: examples/go/node_configuration_example.go
package main

import (
	"encoding/base64"
	"encoding/json"

	"github.com/castai/pulumi-castai/sdk/go/castai" // Assuming main package
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// runNodeConfigurationExample demonstrates node configuration.
// Renamed from main to avoid conflicts.
func runNodeConfigurationExample(ctx *pulumi.Context) error {
	// Placeholder: Assume these IDs/values are obtained from other resources or config
	clusterId := pulumi.String("your-cluster-id")
	subnetIds := pulumi.StringArray{pulumi.String("subnet-12345abc"), pulumi.String("subnet-67890def")} // Example subnet IDs
	instanceProfileArn := pulumi.String("arn:aws:iam::123456789012:instance-profile/YourProfile")       // Example EKS profile ARN
	securityGroupIds := pulumi.StringArray{pulumi.String("sg-12345abc")}                                // Example EKS SG ID

	initScript := `#!/bin/bash
echo "Hello from Pulumi-managed node!"
# Add other initialization commands here
`
	initScriptBase64 := base64.StdEncoding.EncodeToString([]byte(initScript))

	dockerConfigJSON, _ := json.Marshal(map[string]interface{}{
		"insecure-registries":      []string{"my-registry.local:5000"},
		"max-concurrent-downloads": 10,
	})

	kubeletConfigJSON, _ := json.Marshal(map[string]interface{}{
		"registryBurst":   20,
		"registryPullQPS": 10,
	})

	// Create a Node Configuration
	nodeConfig, err := castai.NewNodeConfiguration(ctx, "example-go-node-config", &castai.NodeConfigurationArgs{
		Name:         pulumi.String("default-go-config"),
		ClusterId:    clusterId,
		DiskCpuRatio: pulumi.Int(35),
		// MinDiskSize:      pulumi.Int(133), // Optional
		Subnets:          subnetIds,
		InitScript:       pulumi.String(initScriptBase64),
		DockerConfig:     pulumi.String(string(dockerConfigJSON)),
		KubeletConfig:    pulumi.String(string(kubeletConfigJSON)),
		ContainerRuntime: pulumi.String("containerd"), // Or dockerd
		Tags: pulumi.StringMap{
			"provisioner": pulumi.String("castai-pulumi"),
			"environment": pulumi.String("dev"),
		},
		// Cloud-specific settings (EKS example)
		Eks: &castai.NodeConfigurationEksArgs{
			InstanceProfileArn: instanceProfileArn,
			// DnsClusterIp:       pulumi.String("10.100.0.10"), // Optional
			SecurityGroups: securityGroupIds,
		},
		// Other cloud providers (gke, aks) would have their own blocks
	})
	if err != nil {
		return err
	}

	// Set the created configuration as the default for the cluster
	defaultConfig, err := castai.NewNodeConfigurationDefault(ctx, "example-go-default-config", &castai.NodeConfigurationDefaultArgs{
		ClusterId:       clusterId,
		ConfigurationId: nodeConfig.ID(),
	})
	if err != nil {
		return err
	}

	// Export the configuration ID
	ctx.Export("node_configuration_go_id", nodeConfig.ID())
	// Export the ID of the default setting resource
	ctx.Export("default_config_go_resource_id", defaultConfig.ID())

	return nil
}

/*
// Example of how you might call this from your main function:
func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        if err := runNodeConfigurationExample(ctx); err != nil {
            return err
        }
        // ... potentially call other example functions ...
        return nil
    })
}
*/

File: examples/typescript/node_configuration_example.ts
import * as pulumi from "@pulumi/pulumi";
import * as castai from "@pulumi/castai"; // Use your actual package name
import * as aws from "@pulumi/aws"; // Used for EKS example specifics
import { Buffer } from "buffer"; // Node.js Buffer for base64 encoding

// Placeholder: Assume these IDs/values are obtained from other resources or config
const clusterId = "your-cluster-id";
const subnetIds = ["subnet-12345abc", "subnet-67890def"]; // Example subnet IDs
const instanceProfileArn = "arn:aws:iam::123456789012:instance-profile/YourProfile"; // Example EKS profile ARN
const securityGroupIds = ["sg-12345abc"]; // Example EKS SG ID

const initScript = `#!/bin/bash
echo "Hello from Pulumi-managed node!"
# Add other initialization commands here
`;

const dockerConfig = JSON.stringify({
    "insecure-registries": ["my-registry.local:5000"],
    "max-concurrent-downloads": 10,
});

const kubeletConfig = JSON.stringify({
    "registryBurst": 20,
    "registryPullQPS": 10,
});

// Create a Node Configuration
const nodeConfig = new castai.NodeConfiguration("example-ts-node-config", {
    name: "default-ts-config",
    clusterId: clusterId,
    diskCpuRatio: 35,
    // minDiskSize: 133, // Optional
    subnets: subnetIds,
    initScript: Buffer.from(initScript).toString("base64"),
    dockerConfig: dockerConfig,
    kubeletConfig: kubeletConfig,
    containerRuntime: "containerd", // Or dockerd
    tags: {
        "provisioner": "castai-pulumi",
        "environment": "dev",
    },
    // Cloud-specific settings (EKS example)
    eks: {
        instanceProfileArn: instanceProfileArn,
        // dnsClusterIp: "10.100.0.10", // Optional
        securityGroups: securityGroupIds,
    },
    // Other cloud providers (gke, aks) would have their own blocks
});

// Set the created configuration as the default for the cluster
const defaultConfig = new castai.NodeConfigurationDefault("example-ts-default-config", {
    clusterId: clusterId,
    configurationId: nodeConfig.id,
});

// Export the configuration ID
export const nodeConfigurationId = nodeConfig.id;
// Export the ID of the default setting resource (might not be very useful itself)
export const defaultConfigResourceId = defaultConfig.id; 